provider/rancher.go

```diff
@@ -0,0 +1,458 @@
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/BurntSushi/ty/fun"
+	"github.com/cenk/backoff"
+	"github.com/containous/traefik/job"
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	rancher "github.com/rancher/go-rancher/client"
+	"math"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+)
+
+const (
+	// RancherDefaultWatchTime is the duration of the interval when polling rancher
+	RancherDefaultWatchTime = 15 * time.Second
+)
+
+var _ Provider = (*Rancher)(nil)
+
+// Rancher holds configurations of the Rancher provider.
+type Rancher struct {
+	BaseProvider     `mapstructure:",squash"`
+	Endpoint         string `description:"Rancher server HTTP(S) endpoint."`
+	AccessKey        string `description:"Rancher server access key."`
+	SecretKey        string `description:"Rancher server Secret Key."`
+	ExposedByDefault bool   `description:"Expose Services by default"`
+	Domain           string `description:"Default domain used"`
+}
+
+type rancherData struct {
+	Name       string
+	Labels     map[string]string // List of labels set to container or service
+	Containers []string
+	Health     string
+}
+
+func (r rancherData) String() string {
+	return fmt.Sprintf("{name:%s, labels:%v, containers: %v, health: %s}", r.Name, r.Labels, r.Containers, r.Health)
+}
+
+// Frontend Labels
+func (provider *Rancher) getPassHostHeader(service rancherData) string {
+	if passHostHeader, err := getServiceLabel(service, "traefik.frontend.passHostHeader"); err == nil {
```

Comment:
I really feel this kind of line are duplicated a lot, it's orthogonal to this PR (/me thinking outloud 🤔) but at some point I think we could refactor this to share some common code.